torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -387,6 +394,11 @@ void ProcessGroupGloo::RecvWork::wait() {
   if (exception_) {
     std::rethrow_exception(exception_);
   }
+  return recvCompleted;
+}
+
+void ProcessGroupGloo::RecvWork::abort() {
+  buffer_->abortWaitRecv();
```

Comment:
We should not have to acquire a lock here since this function does not do any writes to class variables, and `buffer_` in gloo should be thread safe.

Reply:
Does this mean we can also avoid holding the lock in `RecvWork::wait` and `SendWork::wait` when we call `buffer_->waitRecv` and `buffer_->waitSend`? We could just acquire the lock when we set `completed_` and `exception_`?

cc @pietern

Reply:
This makes sense to me unless I'm missing something.

Reply:
The semantics for the work is that there is only a single logical user/waiter. The lock is there only to synchronize a background thread that executes the work and marks it as completed, not multiple waiters. For the send/recv work, that background thread is the actual Gloo I/O thread that marks a send/recv as completed, so for this case we could get rid of the locks entirely.

Reply:
Filed https://github.com/pytorch/pytorch/issues/30164 so we don't forget about this.